cmd/http.go

```diff
@@ -57,43 +60,46 @@ func setupHTTPServer(ctx context.Context, cfg *apiServerConfig) (io.Closer, io.C
 	} else {
 		tlsCfg, err := loadTLSConfig(cfg.CertPath, cfg.KeyPath)
 		if err != nil {
-			return nil, nil, err
+			return cancel, err
 		}
 		l, err := tls.Listen("tcp", cfg.Addr, tlsCfg)
 		if err != nil {
-			return nil, nil, errors.Wrap(err, "error setting up listener for pod http server")
+			return cancel, errors.Wrap(err, "error setting up listener for pod http server")
 		}
 
-		mux := http.NewServeMux()
-		vkubelet.AttachPodRoutes(p, mux)
+		if l != nil {
+			mux := http.NewServeMux()
+			vkubelet.AttachPodRoutes(p, mux)
 
-		podS = &http.Server{
-			Handler:   mux,
-			TLSConfig: tlsCfg,
+			s := &http.Server{
+				Handler:   mux,
+				TLSConfig: tlsCfg,
+			}
+			go serveHTTP(ctx, s, l, "pods")
+			closers = append(closers, s)
 		}
-		go serveHTTP(ctx, podS, l, "pods")
+
 	}
 
 	if cfg.MetricsAddr == "" {
 		log.G(ctx).Info("Pod metrics server not setup due to empty metrics address")
 	} else {
 		l, err := net.Listen("tcp", cfg.MetricsAddr)
 		if err != nil {
-			if l != nil {
-				podS.Close()
-			}
-			return nil, nil, errors.Wrap(err, "could not setup listenr for pod metrics http server")
+			cancel()
```

Comment:
Why call `cancel` here and return `cancel`? I'd expect the caller to call `cancel` if this function returns error != nil.